File: submissions/cpp/216.cpp
class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> ret;
        vector<int> temp;
        helper(k, n, ret, temp, 1);
        return ret;
    }

    void helper(int n, int sum, vector<vector<int>> &ret, vector<int> &temp, int index) {
        if (sum == 0 && n == 0) {
            ret.push_back(temp);
            return;
        }
        for (int i = index; i <= 9; ++i) {
            if (sum - i < 0)
                break;
            temp.push_back(i);
            helper(n - 1, sum - i, ret, temp, i + 1);
            temp.pop_back();
        }
    }
};

File: LeetCode/python3/216.py
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        ret = []
        self.Backtrack(k, n, [], ret, 1)
        return ret

    def Backtrack(self, k, sum, combination, ret, m):
        if k == 0:
            if sum == 0:
                ret.append(combination)
            return combination
        for i in range(m, 10):
            if sum - i < 0:
                break
            combination.append(i)
            combination = self.Backtrack(k - 1, sum - i, combination, ret, i + 1)
            combination = combination[:-1]
        return combination